okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java

```diff
@@ -438,37 +450,30 @@ private void onError(ErrorCode errorCode, String moreDetail) {
 
   private void onGoAway(int lastKnownStreamId, Status status) {
     boolean notifyShutdown;
-    ArrayList<OkHttpClientStream> goAwayStreams = new ArrayList<OkHttpClientStream>();
-    List<PendingStream> pendingStreamsCopy;
     synchronized (lock) {
       notifyShutdown = !goAway;
       goAway = true;
       goAwayStatus = status;
-      synchronized (streams) {
-        Iterator<Map.Entry<Integer, OkHttpClientStream>> it = streams.entrySet().iterator();
-        while (it.hasNext()) {
-          Map.Entry<Integer, OkHttpClientStream> entry = it.next();
-          if (entry.getKey() > lastKnownStreamId) {
-            goAwayStreams.add(entry.getValue());
-            it.remove();
-          }
+      Iterator<Map.Entry<Integer, OkHttpClientStream>> it = streams.entrySet().iterator();
+      while (it.hasNext()) {
+        Map.Entry<Integer, OkHttpClientStream> entry = it.next();
+        if (entry.getKey() > lastKnownStreamId) {
+          entry.getValue().transportReportStatus(status, false, new Metadata.Trailers());
+          it.remove();
         }
       }
-      pendingStreamsCopy = pendingStreams;
-      pendingStreams = new LinkedList<PendingStream>();
+
+      for (PendingStream stream : pendingStreams) {
+        stream.clientStream.transportReportStatus(status, true, new Metadata.Trailers());
+        stream.createdFuture.set(null);
+      }
+      pendingStreams.clear();
     }
 
     if (notifyShutdown) {
       listener.transportShutdown();
```

Comment:
FYI: this seems like it will need to be moved into the synchronized block, because another thread could end up calling `stopIfNecessary()` after we leave the synchronized block, and that will call the listener 1) concurrently, which is not permitted, and 2) in the wrong order.

I'm okay if it isn't done in this PR.